docs/guides/slowness.md

```diff
@@ -0,0 +1,230 @@
+  ones and are not present in the original XCTest implementation (only applicable to predicate and class chain locators)
+
+### Solutions
+
+- Reduce the size of the app hierarchy using the [snapshotMaxDepth setting](../reference/settings.md).
+  This might not help if the destination element is deeply nested -
+  it won't be found if the value of this setting is lower than its nesting level.
+- Exclude the `visible` and/or `accessible` attributes from your query
+  (only applicable to predicate and class chain locators). These are
+  custom attributes exclusive to WDA and their calculation is expensive in comparison
+  to other native attributes.
+- Reduce various timeouts similarly to how it's advised in the
+  [Application startup is slow](#pattern-application-startup-is-slow) pattern
+- Fix the source code of the application under test to reduce the amount of accessible elements
+  on a single screen
+- Fix the source code of the application under test to avoid running long operations
+  or animations on the main thread
+
+## Pattern: Various element interactions are slow
+
+### Symptoms
+
+You observe timeouts or unusual slowness while clicking elements or performing other
+actions on them.
+
+### Causes
+
+- The current app hierarchy is too large (e.g. has hundreds of elements). This is a known
+  XCTest limitation.
+- The app is not idling/has active animations
+
+### Solutions
+
+- Reduce various timeouts similarly to how it's advised in the
+  [Application startup is slow](#pattern-application-startup-is-slow) pattern
+- Fix the source code of the application under test to reduce the amount of accessible elements
+  on a single screen
+- Fix the source code of the application under test to avoid running long operations
+  or animations on the main thread
+
+## Pattern: Page source retrieval slow
+
+### Symptoms
+
+You observe timeouts or unusual slowness while retrieving the page of the app.
+
+### Causes
+
+In order to retrieve the page source WDA needs to take a snapshot of the whole accessibility
+hierarchy with all element attributes resolved, which is a time-expensive operation.
+Page source retrieval slowness might be observed if:
+- The current app hierarchy is too large (e.g. has hundreds of elements). This is a known
+  XCTest limitation.
+- The app is not idling/has active animations
+- It takes too long to determine each element's `visible` or `accessible` attributes, which are custom
+  ones and are not present in the original XCTest implementation
+
+### Solutions
+
+- Reduce the size of the app hierarchy using the [snapshotMaxDepth setting](../reference/settings.md).
+  Note that you won't see nested elements in the source tree whose nesting level is lower than
+  the given size.
+- Retrieve the page source without "expensive" attributes using the
+  [mobile: source](../reference/execute-methods.md#mobile-source) method with
+  the appropriate `excludedAttributes` argument value.
```

Comment:
or [`pageSourceExcludedAttributes setting`](../reference/settings.md) for `/source` endpoint.

Reply:
added